cmd/testpeers.go

```diff
@@ -258,46 +460,132 @@ func runSelfTest(ctx context.Context, queuedTestCases []testCaseName, allTestCas
 	}
 }
 
-func peerPingTest(ctx context.Context, _ *testPeersConfig, _ string) testResult {
-	// TODO(kalo): implement real ping
-	select {
-	case <-ctx.Done():
-		return testResult{Verdict: testVerdictFail}
-	default:
-		return testResult{
-			Verdict: testVerdictFail,
-			Error:   errors.New("ping not implemented").Error(),
+func peerPingTest(ctx context.Context, _ *testPeersConfig, tcpNode host.Host, peer p2p.Peer) testResult {
+	tr := testResult{Name: "Ping"}
+
+	ticker := time.NewTicker(3 * time.Second)
+	defer ticker.Stop()
+
+	for ; true; <-ticker.C {
+		select {
+		case <-ctx.Done():
+			tr.Verdict = testVerdictFail
+			tr.Error = errTimeoutInterrupted.Error()
+
+			return tr
+		default:
+			result, err := pingPeerOnce(ctx, tcpNode, peer)
+			if err != nil {
+				tr.Verdict = testVerdictFail
+				tr.Error = err.Error()
+
+				return tr
+			}
+
+			if result.Error != nil {
+				switch {
+				case errors.Is(result.Error, context.DeadlineExceeded):
+					tr.Verdict = testVerdictFail
+					tr.Error = errTimeoutInterrupted.Error()
+
+					return tr
+				case p2p.IsRelayError(result.Error):
+					tr.Verdict = testVerdictFail
+					tr.Error = result.Error.Error()
+
+					return tr
+				default:
+					log.Warn(ctx, "Ping to peer failed, retrying in 3 sec...", nil, z.Str("peer_name", peer.Name))
+					continue
+				}
+			}
+
+			tr.Verdict = testVerdictOk
+
+			return tr
 		}
 	}
+
+	tr.Verdict = testVerdictFail
+	tr.Error = errors.New("no ticker").Error()
+
+	return tr
 }
 
-func peerPingMeasureTest(ctx context.Context, _ *testPeersConfig, _ string) testResult {
-	// TODO(kalo): implement real ping measure
-	s := rand.Int31n(300) + 100 //nolint: gosec // it's only temporary to showcase timeouts
-	time.Sleep(time.Duration(s) * time.Millisecond)
-	select {
-	case <-ctx.Done():
-		return testResult{Verdict: testVerdictFail}
-	default:
-		return testResult{
-			Verdict:     testVerdictFail,
-			Measurement: "10ms",
-			Error:       errors.New("pingMeasure not implemented").Error(),
-		}
+func peerPingMeasureTest(ctx context.Context, _ *testPeersConfig, tcpNode host.Host, peer p2p.Peer) testResult {
+	const thresholdAvg = 200 * time.Millisecond
```

Comment:
I suggest to move all timing constants to the global `const ()` section, for better visibility.